README
distlock

distlock是一个C语言开发的PHP扩展模块，用于实现分布式加锁功能，位于不同服务器上的进程或任务可以通过distlock互斥的访问共享资源。

PHP API说明
----------------------
resource distlock_connect(string host_list, int session_timeout) 
  功能：连接zookeeper服务器。
  参数：
              host_list: zookeeper服务器列表，例如：127.0.0.1:2181,127.0.0.1:2182,127.0.0.1:2183。
        session_timeout: session超时，单位: ms，默认为10000ms。
返回值：连接成功返回与zookeeper的连接句柄，连接失败返回NULL。

resource distlock_init(resource conn_handle, string lock_path)
  功能：创建一把锁并初始化。
  参数：
        conn_handle: 与zookeeper的连接句柄。
          lock_path: 锁的节点路径。比如所有资金账户有关的锁都放在/account-locks目录下，
                     资金账户123对应锁的路径为: /account-locks/123。
返回值：创建锁成功返回一个互斥量mutex，创建失败返回NULL。

int distlock_lock(resource mutex, int timeout)
  参数：timeout表示阻塞超时，单位: 秒，timeout=0表示一直阻塞。
  功能：对锁进行加锁操作，加锁成功返回1，加锁失败返回0。如果该锁已经被锁则会阻塞，一直等到能够加锁成功或者超时。
        需要注意的是一定要检查返回结果来判断是否加锁成功。

int distlock_trylock(resource mutex)
  功能：加锁的非阻塞版本，如果mutex已经被锁则立即返回，不会阻塞。加锁成功返回1，加锁失败返回0。

int distlock_unlock(resource mutex)
  功能：对锁进行解锁操作，解锁成功返回1，解锁失败返回0。
  
int distlock_free(resource mutex)
  功能：销毁锁并释放所有资源。如果一把锁不需要使用了，记得要销毁。
  
int distlock_disconnect(resource conn_handle)
  功能：断开和zookeeper服务器的连接。


API使用例子
----------------------
distlock API的使用可以参考tests/test.php。


ChangeLog
distlock

2013.2.1
 * add:     distlock_trylock() 加锁的非阻塞版本。
 * modify:  distlock_lock(mutex,timeout) 加锁的阻塞版本增加超时机制。
 * fix:     pthread_cond_signal() 修复发送的条件准备好的信号可能丢失的情况。
 
2013.1
 * fix:     修复zoo_lock.c/zkr_lock_lock()中内存泄露bug。
 * fatal:   解决zoo_lock.c/child_floor()中致命错误。  
 * add:     增加加锁的阻塞机制，最初加锁是不会阻塞的。
 * add:     增加了PHP扩展模块，可以在php中使用。
 
2012.12
 * init:    最初版本来源于zookeeper-3.4.5，特别要注意的是zookeeper自带加锁实现有一些致命的错误和缺陷。

tests/c.sh
#!/bin/bash
#test.c
gcc -o test test.c ../zoo_lock.c -g -DTHREADED -I /usr/local/include/zookeeper -I ../ -lzookeeper_mt

#test_mt
gcc -o test_mt test_mt.c -g -DTHREADED -I /usr/local/include/zookeeper -lzookeeper_mt 



tests/test.c
#include <stdio.h>
#include <stdlib.h>
#include "zookeeper.h"
#include "zoo_lock.h"

// gcc -o test test.c ../zoo_lock.c -g -DTHREADED -I /usr/local/include/zookeeper -I ../ -lzookeeper_mt

int main(int argc,char *argv[])
{
  zhandle_t *zh;
  zkr_lock_mutex_t mutex;
  char *lock_path;

  if(argc == 2)
    lock_path = argv[1];
  else
    lock_path = "/account-locks/liuxingzhi";

  zh = zookeeper_init("127.0.0.1:2181,127.0.0.1:2182,127.0.0.1:2183",NULL,300000,0,NULL,0);
  if(!zh) {
    printf("zookeeper_init() error.\n");
    return 0;  
  }

  if(zkr_lock_init(&mutex, zh, lock_path, &ZOO_OPEN_ACL_UNSAFE) != 0) {
    printf("zkr_lock_init() error.\n");
    zookeeper_close(zh);
    return 0;
  }

  if(zkr_lock_lock(&mutex, 5) == 0) { 
    printf("lock failure.\n");
  } else {
    printf("lock ok.\n");
    sleep(10);
  }
  
  if(zkr_lock_unlock(&mutex) != 0) {
    printf("unlock failure.\n");
    zkr_lock_destroy(&mutex);
    zookeeper_close(zh);
    return 0;
  }
  printf("unlock ok.\n");
  
  zkr_lock_destroy(&mutex);
  zookeeper_close(zh);
  return 1;
}


tests/test.php
<?php 
$host_list="127.0.0.1:2181,127.0.0.1:2182,127.0.0.1:2183";
$zh = distlock_connect($host_list,300000);
if($zh == NULL) {
  echo "connect failure\n";
  exit();
}

$lock_path="/account-locks/liuxingzhi";
$mutex = distlock_init($zh,$lock_path);
if($mutex == NULL) {
 echo "lock init failure\n";
 exit();
}

$islock = distlock_lock($mutex, 5);
if($islock == true)
  echo "Get Lock ok\n";
else
  echo "Get Lock failure\n";

sleep(60);

distlock_unlock($mutex);
distlock_free($mutex);
distlock_disconnect($zh);
?>



zoo_lock.c
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#ifdef DLL_EXPORT
#define USE_STATIC_LIB
#endif

#if defined(__CYGWIN__)
#define USE_IPV6
#endif

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <zookeeper_log.h>
#include <time.h>
#include <sys/time.h>
#include <sys/socket.h>
#include <limits.h>
#include <zoo_lock.h>
#include <stdbool.h>
#ifdef HAVE_SYS_UTSNAME_H
#include <sys/utsname.h>
#endif

#ifdef HAVE_GETPWUID_R
#include <pwd.h>
#endif

#define IF_DEBUG(x) if (logLevel==ZOO_LOG_LEVEL_DEBUG) {x;}

 
ZOOAPI int zkr_lock_init(zkr_lock_mutex_t* mutex, zhandle_t* zh,
                      char* path, struct ACL_vector *acl) {
    mutex->zh = zh;
    mutex->path = path;
    mutex->acl = acl;
    mutex->completion = NULL;
    mutex->cbdata = NULL;
    mutex->id = NULL;
    mutex->ownerid = NULL;
    mutex->isOwner = 0;
    pthread_mutex_init(&(mutex->pmutex), NULL);
    pthread_mutex_init(&(mutex->cond_mutex), NULL); // xinz+
    pthread_cond_init(&(mutex->cond),NULL);         // xinz+
    return 0;
}

ZOOAPI int zkr_lock_init_cb(zkr_lock_mutex_t *mutex, zhandle_t* zh,
                         char *path, struct ACL_vector *acl,
                         zkr_lock_completion completion, void* cbdata) {
    mutex->zh = zh;
    mutex->path = path;
    mutex->acl = acl;
    mutex->completion = completion;
    mutex->cbdata = cbdata;
    mutex->isOwner = 0;
    mutex->ownerid = NULL;
    mutex->id = NULL;
    pthread_mutex_init(&(mutex->pmutex), NULL);
    pthread_mutex_init(&(mutex->cond_mutex), NULL); // xinz+
    pthread_cond_init(&(mutex->cond),NULL);         // xinz+
    return 0;
}

/**
 * unlock the mutex
 */
ZOOAPI int zkr_lock_unlock(zkr_lock_mutex_t *mutex) {
    pthread_mutex_lock(&(mutex->pmutex));
    zhandle_t *zh = mutex->zh;
    if (mutex->id != NULL) {
        int len = strlen(mutex->path) + strlen(mutex->id) + 2;
        char buf[len];
        sprintf(buf, "%s/%s", mutex->path, mutex->id);
        int ret = 0;
        int count = 0;
        struct timespec ts;
        ts.tv_sec = 0;
        ts.tv_nsec = (.5)*1000000;
        ret = ZCONNECTIONLOSS;
        while (ret == ZCONNECTIONLOSS && (count < 3)) {
            ret = zoo_delete(zh, buf, -1);
            if (ret == ZCONNECTIONLOSS) {
                LOG_DEBUG(("connectionloss while deleting the node"));
                nanosleep(&ts, 0);
                count++;
            }
        }
        if (ret == ZOK || ret == ZNONODE) {
            zkr_lock_completion completion = mutex->completion;
            if (completion != NULL) {
                completion(1, mutex->cbdata);
            }

            free(mutex->id);
            mutex->id = NULL;
            pthread_mutex_unlock(&(mutex->pmutex));
            return 0;
        }
        LOG_WARN(("not able to connect to server - giving up"));
        pthread_mutex_unlock(&(mutex->pmutex));
        return ZCONNECTIONLOSS;
    }
    pthread_mutex_unlock(&(mutex->pmutex));
    return ZSYSTEMERROR;
}

static void free_String_vector(struct String_vector *v) {
    if (v->data) {
        int32_t i;
        for (i=0; i<v->count; i++) {
            free(v->data[i]);
        }
        free(v->data);
        v->data = 0;
    }
}

static int vstrcmp(const void* str1, const void* str2) {
    const char **a = (const char**)str1;
    const char **b = (const char**) str2;
    return strcmp(strrchr(*a, '-')+1, strrchr(*b, '-')+1); 
} 

static void sort_children(struct String_vector *vector) {
    qsort( vector->data, vector->count, sizeof(char*), &vstrcmp);
}
        
static char* child_floor(char **sorted_data, int len, char *element) {
    char* ret = NULL;
    int i =0;
    for (i=0; i < len; i++) {
        if (strcmp(strrchr(sorted_data[i],'-')+1, strrchr(element,'-')+1) < 0) { //xinz*: strcmp(sorted_data[i], element)
            ret = sorted_data[i];
        }
    }
    return ret;
}

static void lock_watcher_fn(zhandle_t* zh, int type, int state,
                            const char* path, void *watcherCtx) {
    //callback that we registered 
    //should be called
    zkr_lock_mutex_t *mutex = watcherCtx;
    
    zkr_lock_lock((zkr_lock_mutex_t*) watcherCtx, 0);
    
    //xinz+
    pthread_mutex_lock(&(mutex->cond_mutex));
    pthread_cond_signal(&(mutex->cond));
    pthread_mutex_unlock(&(mutex->cond_mutex));
}

/**
 * get the last name of the path
 */
static char* getName(char* str) {
    char* name = strrchr(str, '/');
    if (name == NULL) 
        return NULL;
    return strdup(name + 1);
}

/**
 * just a method to retry get children
 */
static int retry_getchildren(zhandle_t *zh, char* path, struct String_vector *vector, 
                             struct timespec *ts, int retry) {
    int ret = ZCONNECTIONLOSS;
    int count = 0;
    while (ret == ZCONNECTIONLOSS && count < retry) {
        ret = zoo_get_children(zh, path, 0, vector);
        if (ret == ZCONNECTIONLOSS) {
            LOG_DEBUG(("connection loss to the server"));
            nanosleep(ts, 0);
            count++;
        }
    }
    return ret;
}

/** see if our node already exists
 * if it does then we dup the name and 
 * return it
 */
static char* lookupnode(struct String_vector *vector, char *prefix) {
    char *ret = NULL;
    if (vector->data) {
        int i = 0;
        for (i = 0; i < vector->count; i++) {
            char* child = vector->data[i];
            if (strncmp(prefix, child, strlen(prefix)) == 0) {
                ret = strdup(child);
                break;
            }
        }
    }
    return ret;
}

/** retry zoo_wexists
 */
static int retry_zoowexists(zhandle_t *zh, char* path, watcher_fn watcher, void* ctx,
                            struct Stat *stat, struct timespec *ts, int retry) {
    int ret = ZCONNECTIONLOSS;
    int count = 0;
    while (ret == ZCONNECTIONLOSS && count < retry) {
        ret = zoo_wexists(zh, path, watcher, ctx, stat);
        if (ret == ZCONNECTIONLOSS) {
            LOG_DEBUG(("connectionloss while setting watch on my predecessor"));
            nanosleep(ts, 0);
            count++;
        }
    }
    return ret;
}
                        
/**
 * the main code that does the zookeeper leader 
 * election. this code creates its own ephemeral 
 * node on the given path and sees if its the first
 * one on the list and claims to be a leader if and only
 * if its the first one of children in the paretn path
 */
static int zkr_lock_operation(zkr_lock_mutex_t *mutex, struct timespec *ts) {
    zhandle_t *zh = mutex->zh;
    char *path = mutex->path;
    char *id = mutex->id;
    struct Stat stat;
    char* owner_id = NULL;
    int retry = 3;
    do {
        const clientid_t *cid = zoo_client_id(zh);
        // get the session id
        int64_t session = cid->client_id;
        char prefix[30];
        int ret = 0;
#if defined(__x86_64__)
        snprintf(prefix, 30, "x-%016lx-", session);
#else 
        snprintf(prefix, 30, "x-%016llx-", session);
#endif
        struct String_vector vectorst;
        vectorst.data = NULL;
        vectorst.count = 0;
        ret = ZCONNECTIONLOSS;
        ret = retry_getchildren(zh, path, &vectorst, ts, retry);
        if (ret != ZOK)
            return ret;
        struct String_vector *vector = &vectorst;
        mutex->id = lookupnode(vector, prefix);
        free_String_vector(vector);
        if (mutex->id == NULL) {
            int len = strlen(path) + strlen(prefix) + 2;
            char buf[len];
            char retbuf[len+20];
            snprintf(buf, len, "%s/%s", path, prefix);
            ret = ZCONNECTIONLOSS;
            ret = zoo_create(zh, buf, NULL, 0,  mutex->acl, 
                             ZOO_EPHEMERAL|ZOO_SEQUENCE, retbuf, (len+20));
            
            // do not want to retry the create since 
            // we would end up creating more than one child
            if (ret != ZOK) {
                LOG_WARN(("could not create zoo node %s", buf));
                return ret;
            }
            mutex->id = getName(retbuf);
        }
        
        if (mutex->id != NULL) {
            ret = ZCONNECTIONLOSS;
            ret = retry_getchildren(zh, path, vector, ts, retry);
            if (ret != ZOK) {
                LOG_WARN(("could not connect to server"));
                return ret;
            }
            //sort this list
            sort_children(vector);
            owner_id = vector->data[0];
            mutex->ownerid = strdup(owner_id);
            id = mutex->id;
            char* lessthanme = child_floor(vector->data, vector->count, id);
            if (lessthanme != NULL) {
                int flen = strlen(mutex->path) + strlen(lessthanme) + 2;
                char last_child[flen];
                sprintf(last_child, "%s/%s",mutex->path, lessthanme);
                ret = ZCONNECTIONLOSS;
                ret = retry_zoowexists(zh, last_child, &lock_watcher_fn, mutex, 
                                       &stat, ts, retry);
                // cannot watch my predecessor i am giving up
                // we need to be able to watch the predecessor 
                // since if we do not become a leader the others 
                // will keep waiting
                if (ret != ZOK) {
                    free_String_vector(vector);
                    LOG_WARN(("unable to watch my predecessor"));
                    ret = zkr_lock_unlock(mutex);
                    while (ret == 0) {
                        //we have to give up our leadership
                        // since we cannot watch out predecessor
                        ret = zkr_lock_unlock(mutex);
                    }
                    return ret;
                }
                // we are not the owner of the lock
                mutex->isOwner = 0;
            }
            else {
                // this is the case when we are the owner 
                // of the lock
                if (strcmp(mutex->id, owner_id) == 0) {
                    LOG_DEBUG(("got the zoo lock owner - %s", mutex->id));
                    mutex->isOwner = 1;
                    if (mutex->completion != NULL) {
                        mutex->completion(0, mutex->cbdata);
                    }
                    free_String_vector(vector); // xinz+,avoid memory leak.
                    return ZOK;
                }
            }
            free_String_vector(vector);
            return ZOK;
        }
    } while (mutex->id == NULL);
    return ZOK;
}

ZOOAPI int zkr_lock_lock(zkr_lock_mutex_t *mutex, int timeout) {
    pthread_mutex_lock(&(mutex->pmutex));
    zhandle_t *zh = mutex->zh;
    char *path = mutex->path;
    struct Stat stat;
    int exists = zoo_exists(zh, path, 0, &stat);
    int count = 0;
    struct timespec ts;
    ts.tv_sec = 0;
    ts.tv_nsec = (.5)*1000000;
    // retry to see if the path exists and 
    // and create if the path does not exist
    while ((exists == ZCONNECTIONLOSS || exists == ZNONODE) && (count <4)) {
        count++;
        // retry the operation
        if (exists == ZCONNECTIONLOSS) 
            exists = zoo_exists(zh, path, 0, &stat);
        else if (exists == ZNONODE) 
            exists = zoo_create(zh, path, NULL, 0, mutex->acl, 0, NULL, 0);
        nanosleep(&ts, 0);
          
    }

    // need to check if we cannot still access the server 
    int check_retry = ZCONNECTIONLOSS;
    count = 0;
    while (check_retry != ZOK && count <4) {
        check_retry = zkr_lock_operation(mutex, &ts);
        if (check_retry != ZOK) {
            nanosleep(&ts, 0);
            count++;
        }
    }
    pthread_mutex_unlock(&(mutex->pmutex));
    //xinz+begin
    if(timeout != 0) {
      pthread_mutex_lock(&(mutex->cond_mutex));
      if(!zkr_lock_isowner(mutex)) { // if lock failure, block or timeout.
        if(timeout > 0) {
          struct timeval now;
          struct timespec tmout;
          gettimeofday(&now, NULL);
          tmout.tv_sec  = now.tv_sec + timeout;
          tmout.tv_nsec = now.tv_usec * 1000;
          pthread_cond_timedwait(&(mutex->cond),&(mutex->cond_mutex), &tmout);
        } else {
          pthread_cond_wait(&(mutex->cond),&(mutex->cond_mutex));
        }
      }
      pthread_mutex_unlock(&(mutex->cond_mutex));
    } //xinz+end
    return zkr_lock_isowner(mutex);
}

                    
ZOOAPI char* zkr_lock_getpath(zkr_lock_mutex_t *mutex) {
    return mutex->path;
}

ZOOAPI int zkr_lock_isowner(zkr_lock_mutex_t *mutex) {
    return (mutex->id != NULL && mutex->ownerid != NULL 
            && (strcmp(mutex->id, mutex->ownerid) == 0));
}

ZOOAPI char* zkr_lock_getid(zkr_lock_mutex_t *mutex) {
    return mutex->ownerid;
}

ZOOAPI int zkr_lock_destroy(zkr_lock_mutex_t* mutex) {
    if (mutex->id) 
        free(mutex->id);
    mutex->path = NULL;
    mutex->acl = NULL;
    mutex->completion = NULL;
    pthread_mutex_destroy(&(mutex->pmutex));
    pthread_mutex_destroy(&(mutex->cond_mutex)); //xinx+
    pthread_cond_destroy(&(mutex->cond));        //xinz+
    mutex->isOwner = 0;
    if (mutex->ownerid) 
        free(mutex->ownerid);
    return 0;
}



php_distlock.h
/*
  +----------------------------------------------------------------------+
  | PHP Version 5                                                        |
  +----------------------------------------------------------------------+
  | Copyright (c) 1997-2012 The PHP Group                                |
  +----------------------------------------------------------------------+
  | This source file is subject to version 3.01 of the PHP license,      |
  | that is bundled with this package in the file LICENSE, and is        |
  | available through the world-wide-web at the following url:           |
  | http://www.php.net/license/3_01.txt                                  |
  | If you did not receive a copy of the PHP license and are unable to   |
  | obtain it through the world-wide-web, please send a note to          |
  | license@php.net so we can mail you a copy immediately.               |
  +----------------------------------------------------------------------+
  | Author:                                                              |
  +----------------------------------------------------------------------+
*/

/* $Id$ */

#ifndef PHP_DISTLOCK_H
#define PHP_DISTLOCK_H

extern zend_module_entry distlock_module_entry;
#define phpext_distlock_ptr &distlock_module_entry

#ifdef PHP_WIN32
#	define PHP_DISTLOCK_API __declspec(dllexport)
#elif defined(__GNUC__) && __GNUC__ >= 4
#	define PHP_DISTLOCK_API __attribute__ ((visibility("default")))
#else
#	define PHP_DISTLOCK_API
#endif

#ifdef ZTS
#include "TSRM.h"
#endif

PHP_MINIT_FUNCTION(distlock);
PHP_MSHUTDOWN_FUNCTION(distlock);
PHP_RINIT_FUNCTION(distlock);
PHP_RSHUTDOWN_FUNCTION(distlock);
PHP_MINFO_FUNCTION(distlock);

PHP_FUNCTION(distlock_connect);
PHP_FUNCTION(distlock_init);
PHP_FUNCTION(distlock_lock);
PHP_FUNCTION(distlock_trylock);
PHP_FUNCTION(distlock_unlock);
PHP_FUNCTION(distlock_free);
PHP_FUNCTION(distlock_disconnect);

/* 
  	Declare any global variables you may need between the BEGIN
	and END macros here:     

ZEND_BEGIN_MODULE_GLOBALS(distlock)
	long  global_value;
	char *global_string;
ZEND_END_MODULE_GLOBALS(distlock)
*/

/* In every utility function you add that needs to use variables 
   in php_distlock_globals, call TSRMLS_FETCH(); after declaring other 
   variables used by that function, or better yet, pass in TSRMLS_CC
   after the last function argument and declare your utility function
   with TSRMLS_DC after the last declared argument.  Always refer to
   the globals in your function as DISTLOCK_G(variable).  You are 
   encouraged to rename these macros something shorter, see
   examples in any other php module directory.
*/

#ifdef ZTS
#define DISTLOCK_G(v) TSRMG(distlock_globals_id, zend_distlock_globals *, v)
#else
#define DISTLOCK_G(v) (distlock_globals.v)
#endif

#endif	/* PHP_DISTLOCK_H */


/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4 fdm=marker
 * vim<600: noet sw=4 ts=4
 */


php_distlock.c
/*
  +----------------------------------------------------------------------+
  | PHP Version 5                                                        |
  +----------------------------------------------------------------------+
  | Copyright (c) 1997-2012 The PHP Group                                |
  +----------------------------------------------------------------------+
  | This source file is subject to version 3.01 of the PHP license,      |
  | that is bundled with this package in the file LICENSE, and is        |
  | available through the world-wide-web at the following url:           |
  | http://www.php.net/license/3_01.txt                                  |
  | If you did not receive a copy of the PHP license and are unable to   |
  | obtain it through the world-wide-web, please send a note to          |
  | license@php.net so we can mail you a copy immediately.               |
  +----------------------------------------------------------------------+
  | Author:                                                              |
  +----------------------------------------------------------------------+
*/

/* $Id$ */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "php.h"
#include "php_ini.h"
#include "ext/standard/info.h"
#include "php_distlock.h"
#include <zookeeper.h>
#include "zoo_lock.h"

/* If you declare any globals in php_distlock.h uncomment this:
ZEND_DECLARE_MODULE_GLOBALS(distlock)
*/

#define PHP_DISTLOCK_CONN   "distlock connect"
#define PHP_DISTLOCK_MUTEX  "distlock mutex"

/* True global resources - no need for thread safety here */
static int le_connect;
static int le_mutex;

/* {{{ distlock_functions[]
 *
 * Every user visible function must have an entry in distlock_functions[].
 */
const zend_function_entry distlock_functions[] = {
	PHP_FE(distlock_connect,NULL)
	PHP_FE(distlock_init,	NULL)
	PHP_FE(distlock_lock,	NULL)
	PHP_FE(distlock_trylock,	NULL)
	PHP_FE(distlock_unlock,	NULL)
	PHP_FE(distlock_free,	NULL)
	PHP_FE(distlock_disconnect,NULL)
	{NULL, NULL, NULL} 	/* Must be the last line in distlock_functions[], or PHP_FE_END */
};
/* }}} */

/* {{{ distlock_module_entry
 */
zend_module_entry distlock_module_entry = {
#if ZEND_MODULE_API_NO >= 20010901
	STANDARD_MODULE_HEADER,
#endif
	"distlock",
	distlock_functions,
	PHP_MINIT(distlock),
	PHP_MSHUTDOWN(distlock),
	PHP_RINIT(distlock),		/* Replace with NULL if there's nothing to do at request start */
	PHP_RSHUTDOWN(distlock),	/* Replace with NULL if there's nothing to do at request end */
	PHP_MINFO(distlock),
#if ZEND_MODULE_API_NO >= 20010901
	"0.1", /* Replace with version number for your extension */
#endif
	STANDARD_MODULE_PROPERTIES
};
/* }}} */

#ifdef COMPILE_DL_DISTLOCK
ZEND_GET_MODULE(distlock)
#endif

/* {{{ PHP_INI
 */
/* Remove comments and fill if you need to have entries in php.ini
PHP_INI_BEGIN()
    STD_PHP_INI_ENTRY("distlock.global_value",      "42", PHP_INI_ALL, OnUpdateLong, global_value, zend_distlock_globals, distlock_globals)
    STD_PHP_INI_ENTRY("distlock.global_string", "foobar", PHP_INI_ALL, OnUpdateString, global_string, zend_distlock_globals, distlock_globals)
PHP_INI_END()
*/
/* }}} */

/* {{{ php_distlock_init_globals
 */
/* Uncomment this function if you have INI entries
static void php_distlock_init_globals(zend_distlock_globals *distlock_globals)
{
	distlock_globals->global_value = 0;
	distlock_globals->global_string = NULL;
}
*/
/* }}} */

/* {{{ PHP_MINIT_FUNCTION
 */
PHP_MINIT_FUNCTION(distlock)
{
	/* If you have INI entries, uncomment these lines 
	REGISTER_INI_ENTRIES();
	*/
  le_connect = zend_register_list_destructors_ex(NULL, NULL, PHP_DISTLOCK_CONN, module_number);
  le_mutex = zend_register_list_destructors_ex(NULL, NULL, PHP_DISTLOCK_MUTEX, module_number);
	return SUCCESS;
}
/* }}} */

/* {{{ PHP_MSHUTDOWN_FUNCTION
 */
PHP_MSHUTDOWN_FUNCTION(distlock)
{
	/* uncomment this line if you have INI entries
	UNREGISTER_INI_ENTRIES();
	*/
	return SUCCESS;
}
/* }}} */

/* Remove if there's nothing to do at request start */
/* {{{ PHP_RINIT_FUNCTION
 */
PHP_RINIT_FUNCTION(distlock)
{
	return SUCCESS;
}
/* }}} */

/* Remove if there's nothing to do at request end */
/* {{{ PHP_RSHUTDOWN_FUNCTION
 */
PHP_RSHUTDOWN_FUNCTION(distlock)
{
	return SUCCESS;
}
/* }}} */

/* {{{ PHP_MINFO_FUNCTION
 */
PHP_MINFO_FUNCTION(distlock)
{
	php_info_print_table_start();
	php_info_print_table_header(2, "distlock support", "enabled");
	php_info_print_table_end();

	/* Remove comments if you have entries in php.ini
	DISPLAY_INI_ENTRIES();
	*/
}
/* }}} */


/* The previous line is meant for vim and emacs, so it can correctly fold and 
   unfold functions in source code. See the corresponding marks just before 
   function definition, where the functions purpose is also documented. Please 
   follow this convention for the convenience of others editing your code.
*/

/* {{{ proto resource distlock_connect(string host_list, int session_timeout)
    */
PHP_FUNCTION(distlock_connect)
{
	char *host_list = NULL;
	int argc = ZEND_NUM_ARGS();
	int host_list_len;
	long session_timeout;
  zhandle_t *zh;

	if (zend_parse_parameters(argc TSRMLS_CC, "sl", &host_list, &host_list_len, &session_timeout) == FAILURE) 
	  return;
        
  if(session_timeout <= 0)
    session_timeout = 10000;

  zh = zookeeper_init(host_list,NULL,session_timeout,0,NULL,0);
  if(!zh)
    RETURN_NULL();

  ZEND_REGISTER_RESOURCE(return_value, zh, le_connect);
}
/* }}} */

/* {{{ proto resource distlock_init(resource conn_handle, string lock_path)
    */
PHP_FUNCTION(distlock_init)
{
	char *lock_path = NULL;
	int argc = ZEND_NUM_ARGS();
	int conn_handle_id = -1;
	int lock_path_len;
	zval *conn_handle = NULL;
  zhandle_t *zh;
  zkr_lock_mutex_t *mutex;

	if (zend_parse_parameters(argc TSRMLS_CC, "rs", &conn_handle, &lock_path, &lock_path_len) == FAILURE) 
		return;

	if (conn_handle) {
		ZEND_FETCH_RESOURCE(zh, zhandle_t *, &conn_handle, conn_handle_id, PHP_DISTLOCK_CONN, le_connect);
    mutex = (zkr_lock_mutex_t *)emalloc(sizeof(zkr_lock_mutex_t));
    if (zkr_lock_init(mutex, zh, lock_path, &ZOO_OPEN_ACL_UNSAFE) != 0) {
      efree(mutex);
      RETURN_NULL();
    }
    ZEND_REGISTER_RESOURCE(return_value, mutex, le_mutex);
	} else {
    RETURN_NULL();
  }
}
/* }}} */

/* {{{ proto int distlock_lock(resource mutex, int timeout)
    */
PHP_FUNCTION(distlock_lock)
{
	int argc = ZEND_NUM_ARGS();
	int mutex_id = -1;
	int timeout;
	zval *mutex = NULL;
  zkr_lock_mutex_t *lm;

	if (zend_parse_parameters(argc TSRMLS_CC, "rl", &mutex, &timeout) == FAILURE) 
		return;
	
	if(timeout <= 0)
	  timeout = -1;

	if (mutex) {
		ZEND_FETCH_RESOURCE(lm, zkr_lock_mutex_t *, &mutex, mutex_id, PHP_DISTLOCK_MUTEX, le_mutex);
    if(zkr_lock_lock(lm, timeout) == 1)
      RETURN_TRUE;
	} 
	
	RETURN_FALSE;
}
/* }}} */

/* {{{ proto int distlock_trylock(resource mutex)
    */
PHP_FUNCTION(distlock_trylock)
{
	int argc = ZEND_NUM_ARGS();
	int mutex_id = -1;
	zval *mutex = NULL;
  zkr_lock_mutex_t *lm;

	if (zend_parse_parameters(argc TSRMLS_CC, "r", &mutex) == FAILURE) 
		return;
	
	if (mutex) {
		ZEND_FETCH_RESOURCE(lm, zkr_lock_mutex_t *, &mutex, mutex_id, PHP_DISTLOCK_MUTEX, le_mutex);
    if(zkr_lock_lock(lm, 0) == 1)
      RETURN_TRUE;
	} 
	
	RETURN_FALSE;
}
/* }}} */

/* {{{ proto int distlock_unlock(resource mutex)
    */
PHP_FUNCTION(distlock_unlock)
{
	int argc = ZEND_NUM_ARGS();
	int mutex_id = -1;
	zval *mutex = NULL;
  zkr_lock_mutex_t *lm;

	if (zend_parse_parameters(argc TSRMLS_CC, "r", &mutex) == FAILURE) 
		return;

	if (mutex) {
		ZEND_FETCH_RESOURCE(lm, zkr_lock_mutex_t *, &mutex, mutex_id, PHP_DISTLOCK_MUTEX, le_mutex);
    if(zkr_lock_unlock(lm) == 0) 
      RETURN_TRUE;
	} 
        
  RETURN_FALSE;
}
/* }}} */

/* {{{ proto int distlock_free(resource mutex)
    */
PHP_FUNCTION(distlock_free)
{
	int argc = ZEND_NUM_ARGS();
	int mutex_id = -1;
	zval *mutex = NULL;
  zkr_lock_mutex_t *lm;

	if (zend_parse_parameters(argc TSRMLS_CC, "r", &mutex) == FAILURE) 
		return;

	if (mutex) {
		ZEND_FETCH_RESOURCE(lm, zkr_lock_mutex_t *, &mutex, mutex_id, PHP_DISTLOCK_MUTEX, le_mutex);
    zkr_lock_destroy(lm);                
    efree(lm);
    RETURN_TRUE;
	} else
    RETURN_FALSE;
}
/* }}} */

/* {{{ proto int distlock_disconnect(resource conn_handle)
    */
PHP_FUNCTION(distlock_disconnect)
{
	int argc = ZEND_NUM_ARGS();
	int conn_handle_id = -1;
	zval *conn_handle = NULL;
  zhandle_t *zh;

	if (zend_parse_parameters(argc TSRMLS_CC, "r", &conn_handle) == FAILURE) 
		return;

	if (conn_handle) {
		ZEND_FETCH_RESOURCE(zh, zhandle_t *, &conn_handle, conn_handle_id, PHP_distlock_CONN, le_connect);
    zookeeper_close(zh);
    RETURN_TRUE;
	} else
    RETURN_FALSE;
}
/* }}} */

/*
 * Local variables:
 * tab-width: 4
 * c-basic-offset: 4
 * End:
 * vim600: noet sw=4 ts=4 fdm=marker
 * vim<600: noet sw=4 ts=4
 */
